### DISCUSSION TITLE
How to catch Exceptions that are thrown in child controls?

### DISCUSSION THREAD

**Discussion by DevAM-Tools** [AvaloniaUI/Avalonia#14349] [Q&A]
Is there a way to catch an exception that a control throws in one of its parents?
Let's say a click in a button causes an exception and the button does not handle the exception then I would like to catch this exception in the parent container.

What's the best way to do that?

**Comment by maxkatz6** [comment]
There is no way to do so. Exception failed input handler will go directly to the Dispatcher Queue and crash the app.

**Reply by DevAM-Tools** [reply]
Thanks for the instant reply. Would it be an option to introduce some kind of event that is raised in such situations to allow catching these exceptions?

**Reply by StefanKoell** [reply]
Hi!

I'm also interested in handling unhandled exception like in WPF or WinForms. I found that #[PR](https://github.com/AvaloniaUI/Avalonia/pull/13476) which in theory should behave similarly. I'm not sure why it hasn't been merged yet or how I could "test" this implementation.

Can someone enlighten me on:
- is there still a blocking issue causing the PR not being merged?
- how can I test the implementation? I tried to clone the fork from the PR but I'm not successful building the solution or creating packages from that solution.

Thanks!

**Reply by maxkatz6** [reply]
> is there still a blocking issue causing the PR not being merged?

It wasn't yet completed. Some discussions were not resolved.

> how can I test the implementation

I merged master into this branch, it should trigger a new build. After ~30 minutes bot should leave a comment with a build number in that PR.

**Reply by StefanKoell** [reply]
Thanks for the insights. Would be nice to see this capabilities in the main/nightly build soon. I tried to create my own packages from the PR's branch but I failed :(

**Reply by timunie** [reply]
It should have a PR in All feed. Check PR comments for build number

**Reply by StefanKoell** [reply]
Just played around with the PR build and this is a huge improvement. @maxkatz6 is there any way to get this merged into the main branch by either clarifying the remaining questions or providing an "opt-in" for the handling as it is?

**Reply by maxkatz6** [reply]
No specific plans. But if somebody else wants to pick this PR up and complete, it's fine by us.

**Reply by StefanKoell** [reply]
@maxkatz6 Can you help me with this? I really need this kind of functionality and I'm a bit surprised that this isn't a priority for the maintainers. I'm happy to help where I can. I'm just not sure if I can bring something to the table as I have now idea what's missing or what needs to be done. Can you or someone else help me figure out what is needed to get this into Avalonia?

**Reply by DevAM-Tools** [reply]
This feature would be highly welcome for me aswell.

**Reply by maxkatz6** [reply]
> I'm a bit surprised that this isn't a priority for the maintainers

It's not, as we consider global exception handling a non-recommended practice in general. But we are not against having it either. It just needs to be implemented by somebody.

> Can you or someone else help me figure out what is needed to get this into Avalonia

A main idea on how to implement this feature is wrapping each DispatcherOperation in try-catch, redirecting all exceptions to the `Dispatcher.UnhandledException`. You also need to keep in mind, that current Dispatcher code was ported from open-sourced WPF last year, so it should be easier to port UnhandledException handling as well. 

That PR in question already done it, but for it also changes behavior of DispatcherFrame which is not what WPF does. Maybe @ltetak remembers why it's part of DispatcherFrame as well.

That, personally, would be enough for the beginning.
But using this API you might noticed that not every OS event goes through the dispatcher. There are some platform callbacks and input events, that can completely bypass Dispatcher. In short, every platform needs to be double checked, if its input (and not only input) events are raised using DispatcherOperation, but otherwise new handler will also miss them.

**Reply by maxkatz6** [reply]
To visualize what I mean, that's how Button.Click stacktrace looks like on Windows:
![image](https://github.com/AvaloniaUI/Avalonia/assets/3163374/440695b4-c46f-4471-ab26-2cc5c026b52c)

There is no dispatcher operation between WndProc and actual Click handler. Meaning, if any exception was raised, it would kill the app either way, regardless of UnhandledException presence.

**Reply by DevAM-Tools** [reply]
Thanks for the explanation. It sounds like it‘s not that easy.

What I want to achieve: I want my app to be extended by plugins. The plugin can provide a UserControl that will be shown. I want to prevent that an unhandled exception in the plugin crashes the entire app.

Do you habe any advise how to accomplish that?

**Reply by ltetak** [reply]
@maxkatz6 I am absolutely not sure about my solution but it survives the button exception test.
![image](https://github.com/AvaloniaUI/Avalonia/assets/50887170/04a2de63-7a6f-4a26-8cd0-3280e54817ce)

I think there *are* some missing cases though. I haven't worked on this since the PR. I checked the WPF implementation and it looks really complicated to me (or the Avalonia Dispatcher implementation is too different so it is hard to simply use what they have). 
IIRC the problem was when you tried to use Dispatcher->Post method. It did not catch the exception or something like that. So the try/catch may be moved to a better place after all.

**Reply by maxkatz6** [reply]
@ltetak in your case DispatcherFrame will either crash with exception (if event wasn't handled) or silently stop and app will be closed (if event was handled). As there is nothing that restarts DispatcherFrame.
And nothing should restart it, as it's desktop specific. For more general usability, including non-desktop platforms, we should follow WPF and pass events through DispatcherOperation, and handle exceptions only there.

**Reply by StefanKoell** [reply]
@maxkatz6 judging from your previous comments, this is something you or someone on your team would not port this over from WPF, correct?

Since this is basically a showstopper for me, I'm forced to look into it myself. I'm just not sure if I am qualified. I basically know nothing about any platform specific stuff (maybe a bit about Windows) and with my limited experience with Avalonia in general, I'm not sure if I succeed. What's the best move forward? Can someone with more experience in this area help me get this into Avalonia?

**Reply by maxkatz6** [reply]
@StefanKoell these two PRs should be a good start:
https://github.com/AvaloniaUI/Avalonia/pull/14432
https://github.com/AvaloniaUI/Avalonia/pull/14434

Second one might need more work over time

**Reply by StefanKoell** [reply]
@maxkatz6 Thanks a lot for this. I will try out the build. I think it's a good start and should I come across a situation which doesn't work or can be improved, I will get back to you. Again, thank you!

**Reply by StefanKoell** [reply]
@maxkatz6 quick question: are those 2 PRs going to make it into the next release and will this be an opt-in or generally available?

**Reply by maxkatz6** [reply]
Generally available. But can't promise about next release. Depends on whoever from the team have time to review them, and they would consider it a priority, which is probably low.

**Reply by StefanKoell** [reply]
Oh, I see. Keeping my fingers crossed 🤞 that someone can do the review rather sooner than later. 🙏
